### ISSUE TITLE
Function setup(): AT+CFUN=1 has a response time up to 3 minutes. This is ignored in the code, causing AT+CSQ to become invalid

### ISSUE THREAD

**Issue by hdahle** [TelenorStartIoT/arduino-dev-kit-coap#5]
According to the [uBLOX SARA-R4 AT Command Manual](https://www.u-blox.com/en/docs/UBX-17003787) section 5.3, the `AT+CFUN` command has a response time of up to 3 minutes. The code in `setup()` seems to ignore this requirement. This has a serious side-effect: AT-commands that are issued before the `AT+CFUN` has completed will not return valid values.

The code in the `while`loop below suffers from this. What happens seems to be:

1. `AT+CFUN` is issued but the code does not wait for it to complete
2. In the meantime, `status` is defined and initialized to 99
3. The condition for the `while` loop is true so the loop is entered
4. The `AT+CSQ` command is issued (but the `AT+CFUN` has not yet completed)
5. The response to the `AT+CSQ` is therefore empty/invalid, and `status = sub.toInt()` will evaluate to 0
6. We reach the end of the while loop with `status` equal to 0, and we exit the loop

So the loop essentially does nothing.

```
  Serial.print("Modem ready, turn radio on in order to configure it...");
  MODEM.send("AT+CFUN=1");
  MODEM.waitForResponse(2000);
  Serial.println("done.");

  Serial.print("Check attachment until CSQ RSSI indicator is less than 99...");
  String response;
  int status = 99;
  while (status > 98 && status > 0) {
    MODEM.send("AT+CSQ");
    MODEM.waitForResponse(2000, &response);
    String sub = response.substring(6, 8);
    status = sub.toInt(); // Will return 0 if no valid number is found
    delay(1000);
  }
  Serial.println("done.");
```

The fix is to wait for completion of  `AT+CFUN`. This ensures that the `AT+CSQ` command returns valid results. I find it useful to add a print-statement inside the `while` loop, this way you can easily see what happens. A `Serial.println(response)` does the trick.

The fix: Replace the 4 first lines above with

```
  // SARA R4 AT Command Manual Section 5.3: +CFUN response time UP TO 3 MIN
  Serial.print("Modem ready, turn radio on in order to configure it...");
  MODEM.send("AT+CFUN=1");
  do {
    delay(100);
    Serial.print(".");
    MODEM.noop();
  } while (MODEM.waitForResponse(1000) != 1);
  Serial.println("done.");
```
Also, I am puzzled by the condition in the `while()` statement:

````
  int status = 99;
  while (status > 98 && status > 0) {
    ...
  }
````
Why does the code test for `status>0` after testing for `status>98` ? I don't understand. Clearly if  `status` is higher than 98 it will also be higher than 0?

Regardless of the logic behind this, the code should probably be changed: According to  [uBLOX SARA-R4 AT Command Manual](https://www.u-blox.com/en/docs/UBX-17003787) section 7.3, the `AT+CSQ` command responds with

`+CSQ:<signal_power>,<qual>`

Where `signal_power` is 0..31 or 99, where 99 is unknown signal power, and where `qual` is always 99 for the SARA-R4 modules. This means that the condition in the `while` loop could be:

```
int status=99;
while (status>31) {
  ...
}
```

However this is not critical.

**Comment by Pwntus** [comment]
Thank you again @hdahle, these are extremely valuable findings you are making.

The 3 minute response time of `AT+CFUN` seems to be the reason behind a lot of unexplainable issues with the code. I see that we use the proposed do-while loop when issuing `AT+CFUN=15` above, but not for `AT+CFUN=1`. I will make an update with your solution.

The code has gone through various changes and optimizations without proper testing, so I'm not sure about how the logic around signal power has occured. However, I guess it is because the value of `status` will be `0` if parsing the modem response fails. This is the case when `signal_power` is below 10 and only consists of one number.

Let's say the response is: `+CSQ: 4,99`. Here the `signal_power`is a one-number below 10 so the substring will include the following comma:

```cpp
String sub = response.substring(6, 8); // sub will be "4,"
status = sub.toInt(); // Will return 0 since "4," is not a valid number
```

I will update the parsing to be a little bit more "smart" by looking for the comma in the modem response, so that we can avoid parsing error:

```cpp
String response;
int status = 99;
while (status > 31) {
  MODEM.send("AT+CSQ");
  MODEM.waitForResponse(2000, &response);

  // Parse response: +CSQ: <signal_power>,<qual>
  int delimeterIndex = response.indexOf(",");
  if (delimeterIndex != -1) {
    String sub = response.substring(6, delimeterIndex);
    status = sub.toInt(); // Will return 0 if no valid number is found
  }
  delay(1000);
}
```

**Comment by Pwntus** [comment]
I will keep the issue open for a while so that we can follow up the commited solution.

**Comment by hdahle** [comment]
Yep, good point on your example "Let's say the response is: +CSQ: 4,99. Here the signal_power is a one-number below 10 so the substring will include the following comma:"

Your suggestion on adding some smarts to the parsing is a good one.

**Comment by steffennilsen** [comment]
We are approaching one year since last this was discussed. There are many opportunities to improve the reliability of this code, but the core of it should also remain fairly simple to read and understand. If there are any new issues regarding this we should open a new issue.

**Comment by hdahle** [comment]
Well, it's still broken. I agree on making the code simple and readable, but it would be nice if it also worked.

**Comment by hdahle** [comment]
That's ok as long as everybody is aware that:
1. The code is broken
2. Any AT-requests performed after AT+CFUN=1 will return incorrect values
3. If your code relies on results of an AT command then your code will fail
I agree that my proposed fix last year was perhaps not elegant and readable but at least it (1) works and (2) adheres to the manufacturers datasheet

**Comment by steffennilsen** [comment]
We are in the process of handing over the StartIoT maintenance responsibilities. My intent of closing the issue was first and foremost that a year had passed with no activity, if this indeed still is a problem after nearly a year then I should probably look into it then.

I have a codebase that does this in a error prone manner, but its leaning more into raw c programming and away from the beginner friendly approach of Arduino. Still, I am very much for reducing my own future workload by reducing support needs, so I'll dedicate some time going over this.

**Comment by steffennilsen** [comment]
As you have pointed out, AT commands may have unreliable return times, a simple delay is not sufficient and neither is waiting for **any** response on the serial. The correct approach is likely to wait until either `OK`, `ERROR` or command specific errors like `CME ERROR` is passed trough.

**Comment by hdahle** [comment]
Agree.  That's pretty much what I did when I had to rewrite this code in order to make it reliable.

**Comment by steffennilsen** [comment]
I've pulled up my other codebase where I worked on this and it needs some refactoring to hide away more advanced concepts like or implementation details:

* Checking `errno` on int parsing
* Waiting on serialbuffer to be empty is insufficient. Need to check for `OK` / `ERROR`
  * There is no guarantee the modem will send all data and or response before the code runs, flushes the buffer and it is empty
  * The SAMD21(ARM) is so fast it could potentially empty the cache during serial transfer, even AVR based boards could be too fast, 115200 baud isn't particularly fast
* Waiting on modem restart `AT+CFUN=15` is also a corner case, it has an unreliable time execution and my experience is that the `OK` on the serial doesnt always come trough after a reboot.
  * My solution to this issue was to *ping* the modem periodically after reboot with a `AT` command and wait for its `OK` ack

On top of this I experienced, presumably, stack fragmentation on long term running when using varied sized messages. Strings in embedded systems are evil..

In any case, I have code I wrote running that takes care of all of this, but I need to refactor it in a way I can release to the public and keep the appearance of being easy to understand.
